alternator/ttl.cc

```diff
@@ -113,4 +142,494 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+    while (!p->is_exhausted()) {
+        if (abort_source.abort_requested()) {
+            co_return;
+        }
+        // We don't to limit page size in number of rows because there is a
+        // builtin limit of the page's size in bytes. Setting this limit to 1
+        // is useful for debugging the paging code with moderate-size data.
+        uint32_t limit = std::numeric_limits<uint32_t>::max();
+        // FIXME: which timeout?
+        // FIXME: if read times out, need to retry it.
+        std::unique_ptr<cql3::result_set> rs = co_await p->fetch_page(limit, gc_clock::now(), executor::default_timeout());
+        auto rows = rs->rows();
+        auto meta = rs->get_metadata().get_names();
+        std::optional<unsigned> expiration_column;
+        std::vector<unsigned> pk_columns;
+        std::vector<unsigned> ck_columns;
+        for (unsigned i = 0; i < meta.size(); i++) {
+            const cql3::column_specification& col = *meta[i];
+            if (col.name->name() == column_name) {
+                expiration_column = i;
+                break;
+            }
+        }
+        if (!expiration_column) {
+            continue;
+        }
+        // Find the key columns, which we will need for extracting the key
+        // the item to delete.
+        // FIXME: Maybe this can be simplified because the columns are sorted,
+        // we just know that the first columns are always the key columns?
+        for (const column_definition& cdef : s->partition_key_columns()) {
+            for (unsigned i = 0; i < meta.size(); i++) {
+                if (meta[i]->name->name() == cdef.name()) {
+                    pk_columns.push_back(i);
+                }
+            }
+        }
+        for (const column_definition& cdef : s->clustering_key_columns()) {
+            for (unsigned i = 0; i < meta.size(); i++) {
+                if (meta[i]->name->name() == cdef.name()) {
+                    ck_columns.push_back(i);
+                }
+            }
+        }
+        for (const auto& row : rows) {
+            const bytes_opt& cell = row[*expiration_column];
+            if (!cell) {
+                continue;
+            }
+            auto v = meta[*expiration_column]->type->deserialize(*cell);
+            bool expired = false;
+            // FIXME: don't recalculate "now" all the time
+            auto now = gc_clock::now();
+            if (member) {
+                // In this case, the expiration-time attribute we're
+                // looking for is a member in a map, saved serialized
+                // into bytes using Alternator's serialization (basically
+                // a JSON serialized into bytes)
+                // FIXME: is it possible to find a specific member of a map
+                // without iterating through it like we do here and compare
+                // the key?
```

Comment:
well, map is serialized as a vector of pairs, so you can use `find_if`, which is still iterating through the loop, but without open-coding it